cpp/include/cudf/functions.h

```diff
@@ -926,6 +926,40 @@ gdf_error gdf_quantile_aprrox(	gdf_column*  col_in,       //input column with 0
                                 void*        t_erased_res, //type-erased result of same type as column;
                                 gdf_context* ctxt);        //context info
 
+
+/* --------------------------------------------------------------------------*/
+  /**
+   * @brief Given a set of columns, of which a subset of these are defined to be the key columns,
+   * all columns are sorted by the group by columns and returned, along with a column containing
+   * a list of the start indices of each group
+   *
+   * @Param[in]  num_data_cols         The number of columns in the dataset
+   * @Param[in]  data_cols_in          The input columns in the dataset
+   * @Param[in]  num_groupby_cols      The number of key columns by
+   * @Param[in]  groupby_col_indices   The column indices of the input dataset that will be grouped by
+   * @Param[out] data_cols_out         The dataset sorted by the group by columns (needs to be pre-allocated)
+   * @Param[out] group_start_indices   A column containing the starting indices of each group. Indices based off of new sort order. (needs to be pre-allocated)
+   * @Param[in]  ctxt                  The context used to control how nulls are treated in a sort and in group by
+   *   context->flag_nulls_sort_behavior
+   *    GDF_NULL_AS_LARGEST = Nulls are are treated as largest, 
+   *    GDF_NULL_AS_SMALLEST = Nulls are treated as smallest, 
+   *    GDF_NULL_AS_LARGEST_FOR_MULTISORT = Special multi-sort case any row with null is largest>
+   *   context-> flag_groupby_include_nulls 
+   *    GDF_PANDAS_STYLE = Nulls are ignored in group by keys (Pandas style),
+   *    GDF_SQL_STYLE = Nulls are treated as values in group by keys where NULL == NULL (SQL style)>
+   *
+   * @Returns gdf_error with error code on failure, otherwise GDF_SUCESS
+   */
+  /* ----------------------------------------------------------------------------*/
+gdf_error gdf_group_by_without_aggregations(gdf_size_type num_data_cols,
+                                            gdf_column** data_cols_in,
+                                            gdf_size_type num_groupby_cols,
+                                            gdf_index_type * groupby_col_indices,
+                                            gdf_column** data_cols_out,
+                                            gdf_column* group_start_indices,
```

Comment:
It could be, but then we would also need to return the output size which makes it more of a hassle than using a gdf_column, especially when forced to use a C syntax. Remember the size of group_start_indices (which I am renaming to unique_key_indices) is ultimately determined inside the function.

Reply:
Ah, that's a very good point. Hm... I still prefer adding an output parameter like `num_unique_keys` than using a `gdf_column` here. Because internally you've hardcoded the type required for the column inside your function, and that's non-obvious to users of the function, whereas if you just make it a `gdf_index_type[]`, then there's no possible confusion. 

BTW, you're free to make a C++ API for this function. The only caveat is that you also need to provide a C "wrapper" API. You can see what I did for a C++ API in my scatter/gather PR: https://github.com/rapidsai/cudf/blob/eebb5fc81de9b72b6eb8618e4aef0d9db3242bdb/cpp/include/copying.hpp

(note that the PR is still missing the C wrapper API)

Obviously we need to have this conversation in a more visible place, but this is the direction I would like libcudf APIs to go: natively C++ with a thin C wrapper API.